internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -5,61 +5,77 @@
+
+	// PathPrefix is the path to prefix any hashed files when uploading to the artifact bucket.
+	PathPrefix string
+
+	// AssetMappingPath is the path the upload the asset mapping file to.
+	AssetMappingPath string
 }
 
-// Upload uploads static assets to Cloud Storage and returns uploaded file URLs.
-func Upload(fs afero.Fs, source, destination string, opts *UploadOpts) ([]string, error) {
-	matcher := buildCompositeMatchers(buildReincludeMatchers(opts.Reincludes), buildExcludeMatchers(opts.Excludes))
-	var urls []string
-	info, err := fs.Stat(source)
-	if err != nil {
-		return nil, fmt.Errorf("get stat for file %q: %w", source, err)
-	}
-	paths := []string{source}
-	if info.IsDir() {
-		files, err := afero.ReadDir(fs, source)
-		if err != nil {
-			return nil, fmt.Errorf("read directory %q: %w", source, err)
-		}
-		paths = make([]string, len(files))
-		for i, f := range files {
-			paths[i] = filepath.Join(source, f.Name())
+type asset struct {
+	localPath string
+	content   io.Reader
+
+	ArtifactBucketPath string `json:"path"`
+	ServiceBucketPath  string `json:"destPath"`
+}
+
+// UploadFiles hashes each of the files specified in files and uploads
+// them to the path "{PathPrefix}/{hash}". After, it uploads a JSON file
+// to AssetMappingPath that specifies the location of every file in the artifact bucket and it's
+// intended destination path in the service bucket.
+func (u *ArtifactBucketUploader) UploadFiles(files []manifest.FileUpload) error {
+	var assets []asset
+	for _, f := range files {
+		matcher := buildCompositeMatchers(buildReincludeMatchers(f.Reinclude.ToStringSlice()), buildExcludeMatchers(f.Exclude.ToStringSlice()))
+		source := filepath.Join(f.Context, f.Source)
+
+		if err := afero.Walk(u.FS, source, u.walkFn(source, f.Destination, f.Recursive, matcher, &assets)); err != nil {
+			return fmt.Errorf("walk the file tree rooted at %q: %s", source, err)
 		}
-	} else if destination == "" { // only applies to files, not folders
-		destination = source
 	}
-	for _, path := range paths {
-		if err := afero.Walk(fs, path, walkFn(source, destination, opts.Recursive, fs, opts.UploadFn, &urls, matcher)); err != nil {
-			return nil, fmt.Errorf("walk the file tree rooted at %q: %w", source, err)
-		}
+
+	if err := u.uploadAssets(assets); err != nil {
+		return fmt.Errorf("upload assets: %s", err)
 	}
-	return urls, nil
+
+	if err := u.uploadAssetMappingFile(assets); err != nil {
+		return fmt.Errorf("upload asset mapping file: %s", err)
+	}
+	return nil
 }
 
-func walkFn(source, dest string, recursive bool, reader afero.Fs, upload UploadFunc, urlsPtr *[]string, matcher filepathMatcher) filepath.WalkFunc {
+func (u *ArtifactBucketUploader) walkFn(sourcePath, destPath string, recursive bool, matcher filepathMatcher, assets *[]asset) filepath.WalkFunc {
 	return func(path string, info fs.FileInfo, err error) error {
 		if err != nil {
 			return err
 		}
 		if info.IsDir() {
-			if !recursive {
+			if !recursive && path != sourcePath {
```

Comment:
I was also stopped by this line of change and was confused a little. Although after reading the conversation between you & ph it all made sense. I wonder if there is a better way to communicate what you've said here in the code 🤔 (a comment, maybe?)

Reply:
i added a clarification comment, hopefully that helps 🥺